purchases/src/test/java/com/revenuecat/purchases/BillingWrapperTest.kt

```diff
@@ -860,6 +860,48 @@ class BillingWrapperTest {
         assertThat(purchaseType).isEqualTo(PurchaseType.INAPP)
     }
 
+    @Test
+    fun `findPurchaseInPurchaseHistory works`() {
+        setup()
+        val sku = "aPurchase"
+        val purchaseHistoryRecord = mockk<PurchaseHistoryRecord>(relaxed = true).also {
+            every { it.sku } returns sku
+        }
+
+        var recordFound: PurchaseHistoryRecordWrapper? = null
+        wrapper.findPurchaseInPurchaseHistory(BillingClient.SkuType.SUBS, sku) { result, record ->
+            recordFound = record
+        }
+        billingClientPurchaseHistoryListener!!.onPurchaseHistoryResponse(
+            BillingClient.BillingResponseCode.OK.buildResult(),
+            listOf(purchaseHistoryRecord)
+        )
+        assertThat(recordFound).isNotNull
+        assertThat(recordFound!!.purchaseHistoryRecord).isEqualTo(purchaseHistoryRecord)
+    }
+
+    @Test
+    fun `findPurchaseInPurchaseHistory returns null if not found`() {
+        setup()
+        val sku = "aPurchase"
+        val purchaseHistoryRecord = mockk<PurchaseHistoryRecord>(relaxed = true).also {
+            every { it.sku } returns sku + "somethingrandom"
+        }
+
+        var recordFound: PurchaseHistoryRecordWrapper? = null
+        var completionCalled = false
+        wrapper.findPurchaseInPurchaseHistory(BillingClient.SkuType.SUBS, sku) { result, record ->
+            recordFound = record
+            completionCalled = true
+        }
+        billingClientPurchaseHistoryListener!!.onPurchaseHistoryResponse(
+            BillingClient.BillingResponseCode.OK.buildResult(),
+            listOf(purchaseHistoryRecord)
+        )
+        assertThat(completionCalled).isTrue()
```

Comment:
is completion guaranteed to have been called by this line? isn't it async?

Reply:
If you look in line 92:

```
every {
            mockClient.queryPurchaseHistoryAsync(
                any(),
                capture(billingClientPurchaseHistoryListenerSlot)
            )
        } answers {
            billingClientPurchaseHistoryListener = billingClientPurchaseHistoryListenerSlot.captured
        }
```

We capture `billingClientPurchaseHistoryListener` but don't execute it, which means that when the function `findPurchaseInPurchaseHistory` \ we will capture the `PurchaseHistoryResponseListener` passed to `billingClient.queryPurchaseHistoryAsync`. The key here is that `billingClient.queryPurchaseHistoryAsync` is mocked and it doesn't do anything. 

So when doing:

```
billingClientPurchaseHistoryListener!!.onPurchaseHistoryResponse(
            BillingClient.BillingResponseCode.OK.buildResult(),
            listOf(purchaseHistoryRecord)
        )
```

What we basically do is mimicking as if the `queryPurchaseHistoryAsync` would call the listener. So in this piece of code, the lambda will be called with a result, `and` a `purchasesList`, so the content of the lambda will be executed.

```
billingClient.queryPurchaseHistoryAsync(skuType) { result, purchasesList ->
                completion(
                    result,
                    purchasesList?.firstOrNull { sku == it.sku }?.let {
                        PurchaseHistoryRecordWrapper(it, PurchaseType.fromSKUType(skuType))
                    }
                )
            }
```

Reply:
my bad, I had been on the lookout for a mock but I was looking only in the diff, I'd missed the mock that was already there. sorry for the noise!